GitUI/UserControls/RevisionGrid/RevisionGridControl.cs

```diff
@@ -1285,13 +1262,51 @@ void OnRevisionReadCompleted()
                 // This has to happen on the UI thread
                 ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
                 {
+                    IEnumerable<ObjectId> headParents = null;
```

Comment:
The whole handling of `headParents` makes me concerned. `TryGetParents` executes `rev-list` and the yields elements while those are iterated over. The issue here is that we are iterating over this collection multiple times over and over again, and we're performing multiple allocations (i.e., essentially stressing the GC).

First time in `AddArtificialRevisions() -> _gridView.Add(..., headParents)`. 
Then we are calling `headParents.ToList().IndexOf(...)` or `headParents.Skip(...).ToList()`.

...And we're churning through the memory in https://github.com/gitextensions/gitextensions/pull/9948#discussion_r851716047.

We should consider reading `headParents` collection into some kind of a dictionary (where the lookup is O(1)), and don't reallocating past that point.

Reply:
This is executed only if no revision that were to be selected was found, for instance when filtering.
This tries to find the closest parent where to insert artificial and what to select. (If possible, the data is shared.)
headParents/parents could be long list.

parents must be traversed in order in SetToBeSelectedFromParents so a dictionary would not be of help.

However, the implementation has a very minor impact on the performance, should not be many ms.
If you have concern for performance, it is the git-rev-list call that is the issue, it adds 100 ms. I cannot see anything that requires time in this method (some of the datagrid manipulation requires ms though.)
The solution should optimize for a match in one of the first in the list.
It would be a reasonable limitation to just request maximum 10? 100? revisions from Git to limit memory usage.

Note that the previous implementation would fire the git-rev-list in more situations without any reason and risk messing up the selection of what we actually want.
So this is an improvement regardless.
(To insert artificial just before the first visible HEAD parent is nice but not essential, similiar to select something when filtering. This could be dropped even if I would be annoyed as this was a pain to get efficient in the sense to never request more data than required.)

git-rev-list should be cached (except for artificial), that would be 100 ms improvement in the repeat case.